uniffi_bindgen/src/bindings/backend.rs

```diff
@@ -163,3 +163,112 @@ pub trait CodeDeclaration {
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        oracle.find(self)
+    }
+}
+
+impl CodeTypeDispatch for Record {
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        oracle.find(&self.type_())
+    }
+}
+
+impl CodeTypeDispatch for Enum {
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        oracle.find(&self.type_())
+    }
+}
+
+impl CodeTypeDispatch for Error {
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        oracle.find(&self.type_())
+    }
+}
+
+impl CodeTypeDispatch for Object {
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        oracle.find(&self.type_())
+    }
+}
+
+impl CodeTypeDispatch for CallbackInterface {
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        oracle.find(&self.type_())
+    }
+}
+
+impl CodeTypeDispatch for Field {
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        oracle.find(&self.type_())
+    }
+}
+
+impl CodeTypeDispatch for Argument {
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        oracle.find(&self.type_())
+    }
+}
+
+// Needed to handle &&Type and &&&Type values, which we sometimes end up with in the template code
+impl<T, C> CodeTypeDispatch for T
+where
+    T: std::ops::Deref<Target = C>,
+    C: CodeTypeDispatch,
+{
+    fn code_type_impl(&self, oracle: &dyn CodeOracle) -> Box<dyn CodeType> {
+        self.deref().code_type_impl(oracle)
+    }
+}
+
+impl<T: CodeTypeDispatch> CodeType for T {
+    fn type_label(&self, oracle: &dyn CodeOracle) -> String {
+        self.code_type_impl(oracle).type_label(oracle)
+    }
+
+    fn canonical_name(&self, oracle: &dyn CodeOracle) -> String {
+        self.code_type_impl(oracle).canonical_name(oracle)
+    }
+
+    fn literal(&self, oracle: &dyn CodeOracle, literal: &Literal) -> String {
+        self.code_type_impl(oracle).literal(oracle, literal)
+    }
+
+    fn lower(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+        self.code_type_impl(oracle).lower(oracle, nm)
+    }
+
+    fn write(
+        &self,
+        oracle: &dyn CodeOracle,
+        nm: &dyn fmt::Display,
+        target: &dyn fmt::Display,
+    ) -> String {
+        self.code_type_impl(oracle).write(oracle, nm, target)
+    }
+
+    fn lift(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+        self.code_type_impl(oracle).lift(oracle, nm)
+    }
+
+    fn read(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+        self.code_type_impl(oracle).read(oracle, nm)
+    }
+
+    fn helper_code(&self, oracle: &dyn CodeOracle) -> Option<String> {
+        self.code_type_impl(oracle).helper_code(oracle)
+    }
+
+    fn imports(&self, oracle: &dyn CodeOracle) -> Option<Vec<String>> {
+        self.code_type_impl(oracle).imports(oracle)
+    }
+}
```

Comment:
It took me a while to get why this was necessary: so now we can treat `ComponentInterface` structures as `CodeTypes`! Is that right?

Could provide a comment on the `impl`?